### PR TITLE
Bugfix/206 allow detached head state pushes

### PR DESCRIPTION
**Pull request by vados-xenon**
Add support to mirror backend changes for branch pushes https://github.com/UXPin/uxpin-issues/issues/15

### CODE REVIEW

**Comment by vados-xenon** [comment]
So a few things -- the code on this branch works (even in detached mode) but there are two things that I need to look into:

- `husky` (which auto-installs git commit hooks) require `node` @ `>=10.0+`, are we OK with that? We could also downgrade `husky` or maybe find some way to ignore the error.

- There are some errors when you push from a detached head branch -- though *new* buttons do certainly work (and show the detached state, with `--branch` specified, older buttons seem to produce the following error:

```
rollbar.min.js:2 Error: Cannot find the root preset element:
    revisionId: "5dc33261-4b13-40ce-bf71-dfd1afffd54f_0f259f00ce277c1038f33bc17cfe2ca4aecf8b54",
    presetId: "919a5857-f7af-4a56-a01d-55e9d350a6be"
    at Object.t.getPresetOwnerElementTreeItems (engine-44d6dd7169.js:45)
    at Object.t.createTreeItemsForStoredElement (engine-44d6dd7169.js:9)
    at Object.t.getTreeItems (engine-44d6dd7169.js:20)
    at new e (engine-44d6dd7169.js:9)
    at i.n.recreateTree (engine-44d6dd7169.js:9)
    at i.n.handleTreeDefiningPropertyChange (engine-44d6dd7169.js:9)
    at i.set [as revisionId] (engine-44d6dd7169.js:9)
    at engine-44d6dd7169.js:9
    at Map.forEach (<anonymous>)
    at i.t._initSetOfProperties (engine-44d6dd7169.js:9)
r.<computed> @ rollbar.min.js:2

rollbar.min.js:2 Error: Child (id: 327c14cc2f2146e4b2aab9515703831e) not found in Canvas (id: canvas)
    at i.n._addToStoredElements (engine-44d6dd7169.js:9)
    at i.n._updateStoredElements (engine-44d6dd7169.js:9)
    at i.set (engine-44d6dd7169.js:9)
    at engine-44d6dd7169.js:9
    at Map.forEach (<anonymous>)
    at i.t._initSetOfProperties (engine-44d6dd7169.js:9)
    at i.t.initProperties (engine-44d6dd7169.js:9)
    at i.n.initProperties (engine-44d6dd7169.js:9)
    at t.load (engine-44d6dd7169.js:20)
    at e.<anonymous> (uxpin-651b758836.js:9)
r.<computed> @ rollbar.min.js:2
3872720#?id_page=77214872:1 

Access to XMLHttpRequest at 'https://adobeid-na1.services.adobe.com/ims/check/v4/token?client_id=a7940f22f77f4d1db8941f02f302068e&jslClient=a7940f22f77f4d1db8941f02f302068e&jslVersion=18d23dc&jslMethods=1+4+6+7+10+13+18+20+21+24+25+27+32+35+40+41+42+43+46+47+49+57+67' from origin 'https://app.uxpinstage.com' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.
```

**Comment by matthewchigira** [comment]
> * There are some errors when you push from a detached head branch -- though _new_ buttons do certainly work (and show the detached state, with `--branch` specified, older buttons seem to produce the following error:
> 
> ```
> rollbar.min.js:2 Error: Cannot find the root preset element:
>     revisionId: "5dc33261-4b13-40ce-bf71-dfd1afffd54f_0f259f00ce277c1038f33bc17cfe2ca4aecf8b54",
>     presetId: "919a5857-f7af-4a56-a01d-55e9d350a6be"
>     at Object.t.getPresetOwnerElementTreeItems (engine-44d6dd7169.js:45)
>     at Object.t.createTreeItemsForStoredElement (engine-44d6dd7169.js:9)
>     at Object.t.getTreeItems (engine-44d6dd7169.js:20)
>     at new e (engine-44d6dd7169.js:9)
>     at i.n.recreateTree (engine-44d6dd7169.js:9)
>     at i.n.handleTreeDefiningPropertyChange (engine-44d6dd7169.js:9)
>     at i.set [as revisionId] (engine-44d6dd7169.js:9)
>     at engine-44d6dd7169.js:9
>     at Map.forEach (<anonymous>)
>     at i.t._initSetOfProperties (engine-44d6dd7169.js:9)
> r.<computed> @ rollbar.min.js:2
> 
> rollbar.min.js:2 Error: Child (id: 327c14cc2f2146e4b2aab9515703831e) not found in Canvas (id: canvas)
>     at i.n._addToStoredElements (engine-44d6dd7169.js:9)
>     at i.n._updateStoredElements (engine-44d6dd7169.js:9)
>     at i.set (engine-44d6dd7169.js:9)
>     at engine-44d6dd7169.js:9
>     at Map.forEach (<anonymous>)
>     at i.t._initSetOfProperties (engine-44d6dd7169.js:9)
>     at i.t.initProperties (engine-44d6dd7169.js:9)
>     at i.n.initProperties (engine-44d6dd7169.js:9)
>     at t.load (engine-44d6dd7169.js:20)
>     at e.<anonymous> (uxpin-651b758836.js:9)
> r.<computed> @ rollbar.min.js:2
> 3872720#?id_page=77214872:1 
> 
> Access to XMLHttpRequest at 'https://adobeid-na1.services.adobe.com/ims/check/v4/token?client_id=a7940f22f77f4d1db8941f02f302068e&jslClient=a7940f22f77f4d1db8941f02f302068e&jslVersion=18d23dc&jslMethods=1+4+6+7+10+13+18+20+21+24+25+27+32+35+40+41+42+43+46+47+49+57+67' from origin 'https://app.uxpinstage.com' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.
> ```
FYI this is the same error that you get when you [remove a component from your library but that component was already being used in a prototype](https://github.com/UXPin/uxpin-issues/issues/12).

File: packages/uxpin-merge-cli/bin/uxpin-merge [hunk 1]
```diff
@@ -33,6 +33,7 @@ program
   .option('--cwd <path>', 'working directory: path to root of the DS repository')
   .option('--config <path>', 'path to a config file. Default: `./uxpin.config.js`')
   .option('--token <token>', 'auth token for the library. Default: `UXPIN_AUTH_TOKEN` env variable')
+  .option('--branch <branch>', 'branch to use when uploading.')
```
Comment (matthewchigira) [inline packages/uxpin-merge-cli/bin/uxpin-merge:36]: Should we mention here that `master` is used if nothing is specified?
`.option('--branch <branch>', 'branch to use when uploading. Default: master')`

**Comment by vados-xenon** [comment]
@matthewchigira Yeah I figured that might be what it was -- but since multiple commits can live on the same design I thought multiple branches would be able to as well since I started saving these to different places.

If we start doing migrations, then we need to add some logic to maybe *not* do migrations if the other uses are on the same branch (so you can work with two different commits on the same branch), or just replace *every* occurrence. Another thing that I noticed quite late was that we don't actually know *which* project someone is going to want to replace, so I put in all of them for now. This operation might be fairly difficult and is worth some scrutiny as well (what if someone has tens/hundreds of projects using a design system? it's not a hard change database wise but people might be surprised).

File: packages/uxpin-merge-cli/bin/uxpin-merge (hunk 1 above)
Comment (vados-xenon) [reply packages/uxpin-merge-cli/bin/uxpin-merge:36]: Yup -- I'll add that note for sure, that will make it even more clear, I don't think we note that master is the default anywhere as of now.

- [ ] Add note to command line option description regarding branch default being `master`

**Comment by matthewchigira** [comment]
On first glance, both this and the other PR look good. I'm gonna give it a proper test in the morning, and finish the review, but this is definitely looking on the right track. Great work :+1:

**Comment by matthewchigira** [comment]
Hey @vados-xenon, I might be doing something wrong here, but just wanted to share my testing experience:

1. Followed all the [instructions](https://wikiuxpin.atlassian.net/wiki/spaces/UE/pages/1722843137/Backend+change+guide). 

2. Checked out your branch of the CLI and yarn-linked it to a boilerplate repo.

3. Pushed the boilerplate from default state (`master` branch). By the way I really like this new message you get when you push `master` and it tells you that it has set that as the DS branch!

       ✅ Library bundle uploaded successfully!
       ✅ Library metadata uploaded successfully!
       ✅ Projects set to use DS branch [master]!

4. Created branch and commited a change, then tried to push branch (but didn't specify the argument in the CLI):

   `$ uxpin-merge push --uxpin-domain uxpinstage.com`

        ✅ Library bundle uploaded successfully!
        🛑 There was an error while uploading library metadata! Please try again.
        ERROR: 

    I wonder why ERROR is blank...should it be telling us that we have a branch checked out but that we haven't specified `--branch`? Or should it say "Library is up-to-date & current branch is [master]"?

    `$ git checkout master`
    `$ uxpin-merge push --uxpin-domain uxpinstage.com`

         ✅ Library is up-to-date!
         🛈  Projects using this Design System have been updated to use the current branch [master]

    `$ git checkout feature-branch-1`
    `$ uxpin-merge push --uxpin-domain uxpinstage.com --branch feature-branch-1`

        ✅ Library bundle uploaded successfully!
        ✅ Library metadata uploaded successfully!
        ✅ Projects set to use DS branch [feature-branch-1]!

     But the changes _haven't_ come through to the UI in staging. Do we need to rebuild staging? It didn't seem like we had to do this from the instructions, but I noticed that it has been redployed since your review app was created...

**Comment by vados-xenon** [comment]
Thanks for looking at this @matthewchigira I'm going to try that flow and see if I can reproduce, I'm not sure about error but changing the branches *should* do it. The one thing I have changed recently was the page migration, so I'm kind of worried it's that.

**Comment by vados-xenon** [comment]
Found the issue -- it was one of the changes I made trying to get the tests to work:

https://github.com/UXPin/uxpin-merge-tools/pull/217/files#diff-d3612e7b7402864479bbc59643edc630fe827ca294dcea51b17f6cf62b5aba0cR20

I need to update the documentation the environment should be `development`, *NOT* `test`, or stub out the function during tests...

**Comment by matthewchigira** [comment]
> Found the issue -- it was one of the changes I made trying to get the tests in CI to work properly:
> 
> https://github.com/UXPin/uxpin-merge-tools/pull/217/files#diff-d3612e7b7402864479bbc59643edc630fe827ca294dcea51b17f6cf62b5aba0cR20
> 
> I need to update the documentation the environment should be `development`, _NOT_ `test`, or stub out the function during tests...

Ah that makes sense, I should have spotted that. So if I set: `$ export UXPIN_ENV=dev` that should work ok. I'll give it a test.

**Review by matthewchigira — CHANGES_REQUESTED** [review]
Just had another look through this. All looks good to me (apart the final starting-from-branch issue we noticed on Friday)!

Can you also pull from master and accept the checked in changes? As there were some `package.json` and `yarn.lock` updates related to the Windows/WSL issue. Thanks :pray:

**Review by matthewchigira — APPROVED** [review]

**Pull request merged**